src/js/load-pyodide.js

```diff
@@ -398,68 +395,69 @@ export async function loadPackage(names, messageCallback, errorCallback) {
     }
   }
 
+  if (toLoad.size === 0 && toLoadShared.size === 0) {
+    messageCallback("No new packages to load");
+    return;
+  }
+
   const packageNames = [...toLoad.keys(), ...toLoadShared.keys()].join(", ");
+  const releaseLock = await acquirePackageLock();
   try {
     messageCallback(`Loading ${packageNames}`);
-    let scriptPromises = [];
-    const loaded = [];
-    const failed = {};
-
-    useSharedLibraryWasmPlugin();
-    Module.locateFile_packagesToLoad = toLoadShared;
-    for (const [pkg, uri] of toLoadShared) {
-      const pkgname =
-        (Module.packages[pkg] && Module.packages[pkg].name) || pkg;
-      const scriptSrc =
-        uri === DEFAULT_CHANNEL ? `${baseURL}${pkgname}.js` : uri;
-      messageCallback(`Loading ${pkg} from ${scriptSrc}`);
-      scriptPromises.push(
-        loadScript(scriptSrc)
-          .then(() => {
-            loaded.push(pkg);
-            loadedPackages[pkg] = uri;
-          })
-          .catch((e) => {
-            failed[pkg] = e;
-          })
-      );
+    const sharedLibraryPromises = {};
+    const packagePromises = {};
+    for (const name of toLoadShared.keys()) {
+      if (loadedPackages[name]) {
+        // Handle the race condition where the package was loaded between when
+        // we did dependency resolution and when we acquired the lock.
+        toLoadShared.delete(name);
+        continue;
+      }
+      sharedLibraryPromises[name] = downloadPackage(name);
     }
-
-    // We must start waiting for runDependencies *after* all the JS files are
-    // loaded, since the number of runDependencies may happen to equal zero
-    // between package files loading.
-    try {
-      await Promise.all(scriptPromises).then(waitRunDependency);
-    } finally {
-      delete Module.monitorRunDependencies;
+    for (const [name, channel] of toLoad) {
+      if (loadedPackages[name]) {
+        // Handle the race condition where the package was loaded between when
+        // we did dependency resolution and when we acquired the lock.
+        toLoad.delete(name);
+        continue;
+      }
+      if (channel === DEFAULT_CHANNEL) {
+        packagePromises[name] = downloadPackage(name);
+      } else {
+        packagePromises[name] = downloadPackage(channel);
```

Comment:
Very minor: Maybe better to define a variable that's going to be passed to downloadPackage, to avoid repeating this line twice?

Reply:
Yeah this is a bit sloppy. I rearranged the logic a bit, I think it's better now.